### ISSUE TITLE
Type of formals with 'nil' as default value

### ISSUE THREAD

**Issue by benharsh** [chapel-lang/chapel#10381]
What should the following program print?

```chpl
proc foo(x = nil) {
  writeln(x.type:string);
}

foo();
```

At the time this issue was created (770c9985) this program prints "nil". Should it print 'object' instead?

The motivation for this question relates to the following test:
[test/users/shetag/objRefAsFormArg.chpl](https://github.com/chapel-lang/chapel/blob/770c9985899783791deba6488472347614b8d57d/test/users/shetag/objRefAsFormArg.chpl)

This test has a similar formal, but some old constructor magic allows the user to pass an allocated class. Furthermore, the compiler allowed assignment between the field of type 'A' and the formal. If the formal was of type 'object', this would normally not be allowed.

In short, there are at least two questions to answer:
1) Should the type of 'x' in the first example be 'object' or 'nil' ?
2) Should we attempt to recreate the old constructor semantics for this kind of formal?

Associated future:
[test/classes/initializers/phase1/nilDefaultVal.future](https://github.com/chapel-lang/chapel/blob/770c9985899783791deba6488472347614b8d57d/test/classes/initializers/phase1/nilDefaultVal.future)

**Comment by benharsh** [comment]
Ah, good point about ``nil`` vs. ``object``.

> Am I correct that if the formal is declared as : A things will work just fine? 

Yes, that works.

Unless anyone objects, I'm going to remove the objRefAsFormArg.chpl test and the associated future from this issue involving initializers.

**Comment by bradcray** [comment]
Works for me.

**Comment by benharsh** [comment]
I think we can close this now.
